File: Session 1/code/network-java/src/main/java/com/compnet/practical1/A6ProducerOK.java
/*
 * 
 */

package com.compnet.practical1;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author 
 */
public class A6ProducerOK {

    public static void main(String[] args) throws InterruptedException {
        
        BlockingQueue<String> q = new LinkedBlockingQueue<>();
        
        Display d = new Display();
        
        Thread c = consumer(q, d);
        c.setDaemon(true); // NB: the program terminates when all non-daemon thread are terminated
        c.start();

        sleep(500, 0);
        int n = 10;
        for (int i = 0; i < n; i++) {
            sleep(0, (int) (Math.random()*999999));
            producer(q, d).start();
        }

        sleep(5000, 0); // do not finish right away (so we can see if the daemon consumer prints)
    }

    private static Thread consumer(final BlockingQueue<String> q, final Display d) {
        return new Thread(() -> {
            try {
                while (true) {
                    String e = q.take();
                    if ("RENDER".equals(e)) {
                        d.render_clean();
                    } else {
                        String[] parts = e.split("/");
                        d.add_value(Integer.parseInt(parts[0]), Integer.parseInt(parts[1]));
                    }
                }
            } catch (InterruptedException ex) {
                Logger.getLogger(A5ProducerKO.class.getName()).log(Level.SEVERE, null, ex);
            }            
        });
    }
    private static Thread producer(final BlockingQueue<String> q, final Display d) {
        return new Thread(() -> {
            try {
                for (int i = 0; i < 100; i++) {
                    int ind = (int) (Math.random()*20);
                    q.put(ind + "/" + 1);
                    sleep((int) (Math.random()*2), 0);
                    q.put(ind + "/" + -1);
                }
                q.put("RENDER");
            } catch (InterruptedException ex) {
                Logger.getLogger(Display.class.getName()).log(Level.SEVERE, null, ex);
            }
        });
    }

    private static void sleep(long ms, int nanos) {
        try {
            Thread.sleep(ms, nanos);
        } catch (InterruptedException ex) {
            Logger.getLogger(Display.class.getName()).log(Level.SEVERE, null, ex);
        }
    }
}

File: Session 1/code/network-python/A6_producer_ok.py

from threading import Thread
import time
import random
import display
import queue


def main(): # called at the end of the file
    d = display.Display()

    q = queue.Queue()

    c = consumer(d, q)
    c.daemon = True # NB: the program terminates when all non-daemon thread are done
    c.start()

    time.sleep(1)
    n = 10
    for i in range(n):
        time.sleep(random.uniform(0, 0.001))
        producer(q).start()

    time.sleep(5) # do not finish right away (so we can see if the consumer prints)

def producer(q):
    def add_elements():
        for iteration in range(100):
            i = random.randrange(20)
            q.put((i, 1))
            time.sleep(random.uniform(0, 0.002))
            q.put((i, -1))
        q.put("RENDER")

    t = Thread(target=add_elements)
    return t


def consumer(d, q):
    def consume():
        while True:
            e = q.get()
            if e == "RENDER":
                d.render_clean()
            else:
                i, v = e
                d.add_value(i, v)

    t = Thread(target=consume)
    return t


if __name__== "__main__":
  main()
